### ISSUE TITLE
Tags at the examples block are not treated as actual tags in scenario

### ISSUE THREAD

**Issue by williamlabrum** [cucumber/cucumber-jvm#849]
It seems as though tags specified at an examples block are not treated as such when being assigned to a scenario:

```
Scenario Outline: Test Stuff
  Given foo
  When bar
  Then fizz

@myTag
Examples:
  | col a |
  | data  |
```

``` @Before
    public void before(Scenario scenario){
        for(String tag : scenario.getSourceTagNames()){
            System.out.print("Tag: " + tag);
        }
    }
```

This ultimately prevents us from creating Hooks based on tagged Example blocks. Perhaps there is an alternative?

**Comment by aslakhellesoy** [comment]
Sounds like a bug

**Comment by brasmusson** [comment]
@williamlabrum Which version of Cucumber-JVM are you using? I cannot reproduce this problem in Cucumber-JVM v1.2.2. There both hooks tagged with the Examples table tag are executed for the scenarios instantiated from the Example table, and `scenario.getSourceTagName()` in a before hook will include the Examples table tag in the return value.

**Comment by williamlabrum** [comment]
@brasmusson Yes, I am using v1.2.2. Another thing that is worth noting is that I am using java 1.6 (not by choice).

Please feel free to clone the cucumber-debugging project here: https://github.com/williamlabrum/cucumber-debugging to see an example setup. 

__Execution:__ mvn clean test -Dcucumber.options="--tags @myTag"

Here is the output seen by me (no indication of output tag data provided by the BeforeStepDefs class):
![image](https://cloud.githubusercontent.com/assets/11874212/7091495/e90a2cac-df5e-11e4-8fd4-a85b1d5f6340.png)

Also, since updating to v1.2.2, I now see a bunch of that blue text when using Scenario Outlines. Whats with that? Makes me feel like I'm skipping scenarios/steps when looking at the output (even though I'm not).

**Comment by williamlabrum** [comment]
I have created a second test that shows that it works when placing the tag at the Scenario Outline level but not at the example block level.

__Execution__: mvn clean test

![image](https://cloud.githubusercontent.com/assets/11874212/7091617/dbe8d4c8-df5f-11e4-97bc-7d0a9282ea1b.png)

**Comment by brasmusson** [comment]
Ok, @williamlabrum its a bug. But it is not as easy as that the tags on examples tables are totally ignored, they are lost when performing tag filtering. 

If you remove `tags = { "~@in-progress", "~@skip"}` for the `@CucumberOption` annotation of your [runner class](https://github.com/williamlabrum/cucumber-debugging/blob/master/src/test/java/com/williamlabrum/cucumber/TestDriver.java#L20), you will see that the `@myTag` tag will exist on the scenario from the examples table in question. If you change the tag value to `tags = { "@myTag"}` in the `@CucumberOption` you will see that the scenario from the examples table in question will be executed, so when filtering on the tags it will be selected, but then when it is executed, the scenario will no longer have the `@myTag` tag.

I have not tracked down where the bug is located, but as it is related to tag filtering, the possibility is that it is located in the gherkin library, and in that case the bug is likely to remain until the current gherkin library ([gherkin2](https://github.com/cucumber/gherkin)) is replaced by the [gherkin3](https://github.com/cucumber/gherkin3) library (since no new releases of gherkin2 is planned).

**Comment by williamlabrum** [comment]
Got it. Thanks for your feedback and support.

**Comment by AlisonOatley** [comment]
Has this bug been looked at? This still seems to be an issue?

**Comment by dkowis** [comment]
Is there a test for it against gherkin3?

**Comment by AlisonOatley** [comment]
<dependency>
	<groupId>info.cukes</groupId>
	<artifactId>gherkin3</artifactId>
		<version>3.0.0</version>
</dependency>

I am using gherkin3 and am still unable to get the example tags using either a before hook:
	@Before 
	public static void scenarioHook(Scenario sc)
	{
		System.out.println("Tags: " + sc.getSourceTagNames());
	}
This prints just the Scenario Outline's tag
Or in my cucumber formatter class
  @Override
  public void examples(Examples examples)
  {
	  for(Tag tag : examples.getTags())
	  {
		  tags.add(tag.getName());
	  }
	  super.examples(examples);
  }
This returns empty

**Comment by aslakhellesoy** [comment]
Cucumber-JVM doesn't use Gherkin3 yet, so adding it to the classpath has no effect

**Comment by AlisonOatley** [comment]
Ok, I thought that might be the case...

Is there an estimate on when Gherkin3 would be added to Cucumber-JVM?

**Comment by aslakhellesoy** [comment]
I don't do estimates. It will be ready when it's ready.

**Comment by sukeshjain** [comment]
Method examples of FilterFormatter class in Gherkin2 package has been changed to fix this problem. However, new version of Gherkin2 is not released. Can we have a new version of Gherkin2 to solve this issue?

**Comment by greenhammock** [comment]
Hello, I think I encountered the same bug, but the behavior that caught my attention is slightly different. I noticed that tags attached to 'Examples' section in 'Scenario Outline' do not appear in the JSON report. I hope the fix you are talking about in this issue will also result in correct JSON output.

**Comment by brasmusson** [comment]
AFAIK, fixing this before v2.0 would mean to make changes to code that will be thrown away by #1035 (use Gherkin4) or it might even require a new release for Gherkin2, so I would prefer to not try to fix this before v2.0.

**Comment by brasmusson** [comment]
Now when #1035 has been merged, the 2.0.0-SNAPSHOT version from the master branch (with the groupId io.cucumber), treats tags on the example blocks correctly.
